mailpoet/lib/Validator/Validator.php

```diff
@@ -0,0 +1,190 @@
+   * It only coerces int to float (e.g. 5 to 5.0).
+   *
+   * @param mixed $value
+   * @return mixed
+   */
+  public function validate(Schema $schema, $value, string $paramName = 'value') {
+    $result = $this->validateAndSanitizeValueFromSchema($value, $schema->toArray(), $paramName);
+    if ($result instanceof WP_Error) {
+      throw ValidationException::createFromWpError($result);
+    }
+    return $result;
+  }
+
+  /**
+   * @param mixed $value
+   * @param array $schema
+   * @param string $paramName
+   * @return mixed|WP_Error
+   */
+  private function validateAndSanitizeValueFromSchema($value, array $schema, string $paramName) {
+    // nullable
+    $fullType = $schema['type'] ?? null;
+    if (is_array($fullType) && in_array('null', $fullType, true) && $value === null) {
+      return null;
+    }
+
+    // anyOf, oneOf
+    if (isset($schema['anyOf'])) {
+      return $this->validateAndSanitizeAnyOf($value, $schema, $paramName);
+    } elseif (isset($schema['oneOf'])) {
+      return $this->validateAndSanitizeOneOf($value, $schema, $paramName);
+    }
+
+    // make types strict
+    $type = is_array($fullType) ? $fullType[0] : $fullType;
+    switch ($type) {
+      case 'number':
+        if (!is_float($value) && !is_int($value)) {
+          return $this->getTypeError($paramName, $fullType);
+        }
+        break;
+      case 'integer':
+        if (!is_int($value)) {
+          return $this->getTypeError($paramName, $fullType);
+        }
+        break;
+      case 'boolean':
+        if (!is_bool($value)) {
+          return $this->getTypeError($paramName, $fullType);
+        }
+        break;
+      case 'array':
+        if (!is_array($value)) {
+          return $this->getTypeError($paramName, $fullType);
+        }
+
+        if (isset($schema['items'])) {
+          foreach ($value as $i => $v) {
+            $result = $this->validateAndSanitizeValueFromSchema($v, $schema['items'], $paramName . '[' . $i . ']');
+            if ($this->wp->isWpError($result)) {
+              return $result;
+            }
+          }
+        }
+        break;
+      case 'object':
+        if (!is_array($value) && !$value instanceof stdClass && !$value instanceof JsonSerializable) {
+          return $this->getTypeError($paramName, $fullType);
+        }
+
+        // ensure string keys
+        $value = (array)($value instanceof JsonSerializable ? $value->jsonSerialize() : $value);
+        if (count(array_filter(array_keys($value), 'is_string')) !== count($value)) {
+          return $this->getTypeError($paramName, $fullType);
+        }
+
+        // validate object properties
+        foreach ($value as $k => $v) {
+          if (isset($schema['properties'][$k])) {
+            $result = $this->validateAndSanitizeValueFromSchema($v, $schema['properties'][$k], $paramName . '[' . $k . ']');
+            if ($this->wp->isWpError($result)) {
+              return $result;
+            }
```

Comment:
If I'm not mistaken we end validation with the first error here. What about collecting errors from all properties and combining  them using `$this->wp->restGetCombiningOperationError($value, $paramName, $errors);` like we do in `validateAndSanitizeAnyOf`? I think it might be handy to get all errors at once.

Reply:
This is only mirroring the behavior of WordPress. I'll add some comments on which functions are mirrored.

Reply:
Ok. Let's keep the behavior similar to what WP does.